lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -58,13 +93,72 @@ FModuleOp LowerLayersPass::buildNewModule(OpBuilder &builder, Location location,
   return newModule;
 }
 
-/// Process a module to remove any layer blocks it has.
-void LowerLayersPass::runOnModule(FModuleOp moduleOp) {
+void LowerLayersPass::removeLayersFromValue(Value value) {
+  auto type = dyn_cast<RefType>(value.getType());
+  if (!type || !type.getLayer())
+    return;
+  value.setType(type.removeLayer());
+}
+
+void LowerLayersPass::removeLayersFromPorts(FModuleLike moduleLike) {
+  auto oldTypeAttrs = moduleLike.getPortTypesAttr();
+  SmallVector<Attribute> newTypeAttrs;
+  newTypeAttrs.reserve(oldTypeAttrs.size());
+  bool changed = false;
+
+  for (auto typeAttr : oldTypeAttrs.getAsRange<TypeAttr>()) {
+    if (auto refType = dyn_cast<RefType>(typeAttr.getValue())) {
+      if (refType.getLayer()) {
+        typeAttr = TypeAttr::get(refType.removeLayer());
+        changed = true;
+      }
+    }
+    newTypeAttrs.push_back(typeAttr);
+  }
+
+  if (!changed)
```

Comment:
Ah, change tracking saves from dead attr update (maybe not so bad but cheaper to not!) as well as an unnecessary walk of the arguments and their types.  Cool!